internal/backend/runtime/omni/controllers/reporting/reporting.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package reporting
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/stripe/stripe-go/v74"
+	"github.com/stripe/stripe-go/v74/subscriptionitem"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+)
+
+var (
+	// DebounceDuration defines how long to wait after the last event before processing updates.
+	DebounceDuration = 10 * time.Minute
+)
+
+// StripeMetricsReporterControllerName is the name of the StripeMetricsReporterController.
+const StripeMetricsReporterControllerName = "StripeMetricsReporterController"
+
+// StripeMetricsReporterController ...
+type StripeMetricsReporterController struct {
+	generic.NamedController
+}
+
+// NewStripeMetricsReporterController initializes StripeMetricsReporterController.
+func NewStripeMetricsReporterController() *StripeMetricsReporterController {
+	return &StripeMetricsReporterController{
+		NamedController: generic.NamedController{
+			ControllerName: StripeMetricsReporterControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *StripeMetricsReporterController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: resources.DefaultNamespace,
+			Type:      omni.MachineStatusMetricsType,
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Outputs() []controller.Output {
+	return nil
+}
+
+// Run implements controller.Controller interface.
+func (ctrl *StripeMetricsReporterController) Run(ctx context.Context, r controller.Runtime, _ *zap.Logger) error {
+	// Set your Stripe API key from the environment variable.
+	stripe.Key = os.Getenv("STRIPE_API_KEY")
+	if stripe.Key == "" {
+		return fmt.Errorf("STRIPE_API_KEY environment variable not set")
+	}
+
+	// Retrieve Subscription Item ID from environment variables or configuration.
+	subscriptionItemID := os.Getenv("STRIPE_SUBSCRIPTION_ITEM_ID")
+	if subscriptionItemID == "" {
+		return fmt.Errorf("STRIPE_SUBSCRIPTION_ITEM_ID environment variable not set")
+	}
```

Comment:
Should we instead do this outside of the controller and crash the whole instance if these are not set? Or this controller crash-looping but functional Omni is more preferable?

Reply:
I could see an argument for either way. I want to lean towards functional Omni but that has it's downsides. I will let others chime in.

Reply:
I'll add a little context as to why I made these environment variables too: With Kubernetes we can at least reference these from a secret.

Reply:
I'd suggest to pass stripe configs to the `NewStripeMetricsReporterController` methods as the arguments.
It will be more obvious that the controller needs some params.